File: backend/src/main/java/com/radar/prospection/claude/MissionAnalysisService.java
package com.radar.prospection.claude;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.radar.prospection.config.RadarProperties;
import com.radar.prospection.domain.Mission;
import com.radar.prospection.domain.MissionStatus;
import com.radar.prospection.domain.Signal;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class MissionAnalysisService {

    private final ClaudeService claude;
    private final RadarProperties properties;

    private static final String MISSION_SYSTEM_PROMPT = """
        Tu es un assistant spécialisé pour les freelances Java Tech Lead.
        Tu analyses des offres de missions freelance et tu retournes UNIQUEMENT un JSON valide, sans aucun texte autour.

        Ton rôle :
        1. Évaluer le fit de la mission avec le profil du freelance (score 0-100)
        2. Identifier qui contacter en priorité (décideur vs RH vs manager)
        3. Détecter les éléments différenciants pour personnaliser l'approche
        4. Identifier les signaux d'urgence (mission ouverte depuis longtemps, langage urgent)
        """;

    private static final String SIGNAL_SYSTEM_PROMPT = """
        Tu es un expert en prospection B2B pour freelances tech.
        Tu analyses des signaux du marché caché et tu retournes UNIQUEMENT un JSON valide, sans texte autour.

        Pour chaque signal, évalue :
        1. Le score d'opportunité (0-100) pour un Tech Lead Java freelance
        2. Pourquoi c'est une opportunité
        3. L'action recommandée et le timing
        """;

    public Mission analyze(Mission mission) {
        try {
            RadarProperties.Profile profile = properties.getProfile();
            String profileContext = buildProfileContext(profile);
            String missionContext = buildMissionContext(mission);

            String prompt = profileContext + "\n\n" + missionContext + "\n\n" +
                """
                Retourne UNIQUEMENT ce JSON :
                {
                  "fitScore": <0-100>,
                  "fitSummary": "<3-4 phrases : pourquoi c'est un bon/mauvais fit, points forts, points faibles>",
                  "decisionMakerHint": "<qui contacter : rôle, comment l'approcher, angle d'accroche>",
                  "urgencySignals": ["<signal1>", "<signal2>"],
                  "recommendedAction": "<APPLY|RESEARCH_MORE|SKIP>"
                }
                """;

            MissionAnalysisResult result = claude.completeAsJson(
                MISSION_SYSTEM_PROMPT, prompt, MissionAnalysisResult.class
            );

            mission.setFitScore(result.getFitScore());
            mission.setFitSummary(result.getFitSummary());
            mission.setDecisionMakerHint(result.getDecisionMakerHint());
            mission.setStatus(MissionStatus.ANALYZED);
            mission.setAnalyzedAt(LocalDateTime.now());

            log.info("[Analyse] '{}' → score: {}", mission.getTitle(), result.getFitScore());

        } catch (Exception e) {
            log.warn("[Analyse] Echec pour '{}': {}", mission.getTitle(), e.getMessage());
            mission.setFitSummary("Analyse impossible: " + e.getMessage());
            mission.setStatus(MissionStatus.ANALYZED);
        }

        return mission;
    }

    public Signal analyzeSignal(Signal signal) {
        try {
            String prompt = """
                Signal détecté :
                Type : %s
                Entreprise : %s
                Titre : %s
                Description : %s

                Retourne UNIQUEMENT ce JSON :
                {
                  "opportunityScore": <0-100>,
                  "opportunityReason": "<pourquoi c'est une opportunité pour un Tech Lead Java freelance>",
                  "suggestedAction": "<action concrète à faire dans les 48h>",
                  "bestContactRole": "<CTO|VP_ENG|TEAM_LEAD|HR|FOUNDER>",
                  "timingAdvice": "<quand contacter et pourquoi>"
                }
                """.formatted(
                    signal.getType(),
                    signal.getCompany(),
                    signal.getTitle(),
                    signal.getDescription()
                );

            SignalAnalysisResult result = claude.completeAsJson(
                SIGNAL_SYSTEM_PROMPT, prompt, SignalAnalysisResult.class
            );

            signal.setOpportunityScore(result.getOpportunityScore());
            signal.setOpportunityReason(result.getOpportunityReason());
            signal.setSuggestedAction(result.getSuggestedAction());

            log.info("[Signal] '{}' → score: {}", signal.getCompany(), result.getOpportunityScore());

        } catch (Exception e) {
            log.warn("[Signal] Echec pour '{}': {}", signal.getCompany(), e.getMessage());
        }

        return signal;
    }

    private String buildProfileContext(RadarProperties.Profile profile) {
        return """
            PROFIL FREELANCE :
            Compétences : %s
            TJM cible : %d - %d €/jour
            Préférence télétravail : %s
            Localisation : %s
            """.formatted(
                String.join(", ", profile.getSkills()),
                profile.getTjmMin(),
                profile.getTjmMax(),
                profile.getRemotePreference(),
                profile.getLocation()
            );
    }

    private String buildMissionContext(Mission mission) {
        return """
            MISSION À ANALYSER :
            Titre : %s
            Description : %s
            Skills requis : %s
            TJM proposé : %d - %d €/jour
            Durée : %s
            Localisation : %s
            Télétravail : %s
            Source : %s
            """.formatted(
                mission.getTitle(),
                mission.getDescription() != null ? mission.getDescription() : "Non précisée",
                mission.getSkills() != null ? String.join(", ", mission.getSkills()) : "Non précisés",
                mission.getTjmMin() != null ? mission.getTjmMin() : 0,
                mission.getTjmMax() != null ? mission.getTjmMax() : 0,
                mission.getDuration() != null ? mission.getDuration() : "Non précisée",
                mission.getLocation() != null ? mission.getLocation() : "Non précisée",
                Boolean.TRUE.equals(mission.getRemote()) ? "Oui" : "Non/Hybride",
                mission.getSource()
            );
    }

    @Data
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class MissionAnalysisResult {
        private int fitScore;
        private String fitSummary;
        private String decisionMakerHint;
        private List<String> urgencySignals;
        private String recommendedAction;
    }

    @Data
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class SignalAnalysisResult {
        private int opportunityScore;
        private String opportunityReason;
        private String suggestedAction;
        private String bestContactRole;
        private String timingAdvice;
    }
}

File: backend/src/main/kotlin/com/radar/prospection/claude/MissionAnalysisService.kt
package com.radar.prospection.claude

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.radar.prospection.config.RadarProperties
import com.radar.prospection.domain.Mission
import com.radar.prospection.domain.MissionStatus
import com.radar.prospection.domain.Signal
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class MissionAnalysisService(
    private val claude: ClaudeService,
    private val properties: RadarProperties
) {
    private val log = LoggerFactory.getLogger(javaClass)

    companion object {
        private const val MISSION_SYSTEM_PROMPT = """
            Tu es un assistant spécialisé pour les freelances Java Tech Lead.
            Tu analyses des offres de missions freelance et tu retournes UNIQUEMENT un JSON valide, sans aucun texte autour.

            Ton rôle :
            1. Évaluer le fit de la mission avec le profil du freelance (score 0-100)
            2. Identifier qui contacter en priorité (décideur vs RH vs manager)
            3. Détecter les éléments différenciants pour personnaliser l'approche
            4. Identifier les signaux d'urgence (mission ouverte depuis longtemps, langage urgent)
            """

        private const val SIGNAL_SYSTEM_PROMPT = """
            Tu es un expert en prospection B2B pour freelances tech.
            Tu analyses des signaux du marché caché et tu retournes UNIQUEMENT un JSON valide, sans texte autour.

            Pour chaque signal, évalue :
            1. Le score d'opportunité (0-100) pour un Tech Lead Java freelance
            2. Pourquoi c'est une opportunité
            3. L'action recommandée et le timing
            """
    }

    fun analyze(mission: Mission): Mission {
        try {
            val profile = properties.profile
            val prompt = buildProfileContext(profile) + "\n\n" + buildMissionContext(mission) + "\n\n" + """
                Retourne UNIQUEMENT ce JSON :
                {
                  "fitScore": <0-100>,
                  "fitSummary": "<3-4 phrases : pourquoi c'est un bon/mauvais fit, points forts, points faibles>",
                  "decisionMakerHint": "<qui contacter : rôle, comment l'approcher, angle d'accroche>",
                  "urgencySignals": ["<signal1>", "<signal2>"],
                  "recommendedAction": "<APPLY|RESEARCH_MORE|SKIP>"
                }
                """

            val result = claude.completeAsJson(MISSION_SYSTEM_PROMPT, prompt, MissionAnalysisResult::class.java)

            mission.fitScore = result.fitScore
            mission.fitSummary = result.fitSummary
            mission.decisionMakerHint = result.decisionMakerHint
            mission.status = MissionStatus.ANALYZED
            mission.analyzedAt = LocalDateTime.now()

            log.info("[Analyse] '{}' → score: {}", mission.title, result.fitScore)

        } catch (e: Exception) {
            log.warn("[Analyse] Echec pour '{}': {}", mission.title, e.message)
            mission.fitSummary = "Analyse impossible: ${e.message}"
            mission.status = MissionStatus.ANALYZED
        }

        return mission
    }

    fun analyzeSignal(signal: Signal): Signal {
        try {
            val prompt = """
                Signal détecté :
                Type : ${signal.type}
                Entreprise : ${signal.company}
                Titre : ${signal.title}
                Description : ${signal.description}

                Retourne UNIQUEMENT ce JSON :
                {
                  "opportunityScore": <0-100>,
                  "opportunityReason": "<pourquoi c'est une opportunité pour un Tech Lead Java freelance>",
                  "suggestedAction": "<action concrète à faire dans les 48h>",
                  "bestContactRole": "<CTO|VP_ENG|TEAM_LEAD|HR|FOUNDER>",
                  "timingAdvice": "<quand contacter et pourquoi>"
                }
                """

            val result = claude.completeAsJson(SIGNAL_SYSTEM_PROMPT, prompt, SignalAnalysisResult::class.java)

            signal.opportunityScore = result.opportunityScore
            signal.opportunityReason = result.opportunityReason
            signal.suggestedAction = result.suggestedAction

            log.info("[Signal] '{}' → score: {}", signal.company, result.opportunityScore)

        } catch (e: Exception) {
            log.warn("[Signal] Echec pour '{}': {}", signal.company, e.message)
        }

        return signal
    }

    private fun buildProfileContext(profile: RadarProperties.Profile) = """
        PROFIL FREELANCE :
        Compétences : ${profile.skills.joinToString(", ")}
        TJM cible : ${profile.tjmMin} - ${profile.tjmMax} €/jour
        Préférence télétravail : ${profile.remotePreference}
        Localisation : ${profile.location}
        """

    private fun buildMissionContext(mission: Mission) = """
        MISSION À ANALYSER :
        Titre : ${mission.title}
        Description : ${mission.description ?: "Non précisée"}
        Skills requis : ${mission.skills.joinToString(", ").ifEmpty { "Non précisés" }}
        TJM proposé : ${mission.tjmMin ?: 0} - ${mission.tjmMax ?: 0} €/jour
        Durée : ${mission.duration ?: "Non précisée"}
        Localisation : ${mission.location ?: "Non précisée"}
        Télétravail : ${if (mission.remote == true) "Oui" else "Non/Hybride"}
        Source : ${mission.source}
        """

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class MissionAnalysisResult(
        val fitScore: Int = 0,
        val fitSummary: String? = null,
        val decisionMakerHint: String? = null,
        val urgencySignals: List<String> = emptyList(),
        val recommendedAction: String? = null
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class SignalAnalysisResult(
        val opportunityScore: Int = 0,
        val opportunityReason: String? = null,
        val suggestedAction: String? = null,
        val bestContactRole: String? = null,
        val timingAdvice: String? = null
    )
}
